proxy/match.go

```diff
@@ -0,0 +1,71 @@
+package proxy
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/pomerium/pomerium/internal/atomicutil"
+	"github.com/pomerium/pomerium/internal/urlutil"
+)
+
+type RouteMatcher interface {
+	GetIdentityProviderForRequest(r *http.Request) (string, error)
+}
+
+type RouteMatcherFunc func(r *http.Request) (string, error)
+
+func (f RouteMatcherFunc) GetIdentityProviderForRequest(r *http.Request) (string, error) {
+	return f(r)
+}
+
+func newCheckRouteAPIMatcher(state *atomicutil.Value[*proxyState]) RouteMatcher {
+	return &checkRouteAPIMatcher{state: state}
+}
+
+type checkRouteAPIMatcher struct {
+	state *atomicutil.Value[*proxyState]
+}
+
+func (l *checkRouteAPIMatcher) GetIdentityProviderForRequest(r *http.Request) (string, error) {
+	requestURL := urlutil.NewSignedURL(
+		l.state.Load().cookieSecret,
+		urlutil.GetAbsoluteURL(r).ResolveReference(&url.URL{
+			Path: r.FormValue(urlutil.QueryRequestPath),
+		}),
+	)
+
+	ctx, ca := context.WithTimeout(r.Context(), 10*time.Second)
+	defer ca()
+	req, _ := http.NewRequestWithContext(ctx, r.Method, requestURL.String(), nil)
+	req.Header.Set("X-Pomerium-Check-Route", "1")
+	req.Header.Set("Accept", "application/json")
+	resp, err := http.DefaultTransport.RoundTrip(req)
```

Comment:
what if there's a client certificate requirement for a route?

Reply:
That wouldn't affect the route matching, would it?

Reply:
wouldn't request fail on TLS handshake, before route matching can happen?

Reply:
oh, good point. I'll look into this